File: src/main/java/com/tort/mudai/mapper/Location.java
package com.tort.mudai.mapper;

import com.sun.xml.internal.bind.v2.util.QNameMap;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class Location {
    private String _title;
    private Map<String, Location> _directions = new HashMap<String, Location>();
    private String _waterSource;
    private boolean _isShop = false;
    private boolean _isTavern = false;
    private String _desc;
    private Set<Directions> _availableExits;
    private Set _knownMobs = new HashSet();

    public String getTitle() {
        return _title;
    }

    public void setTitle(final String locationTitle) {
        _title = locationTitle;
    }

    public Location getByDirection(final String direction) {
        return _directions.get(direction);
    }

    public void addDirection(final String direction, final Location location) {
        _directions.put(direction, location);
    }

    public Set<String> getDirections() {
        return _directions.keySet();
    }

    public String getWaterSource() {
        return _waterSource;
    }

    public void setWaterSource(final String waterSource) {
        _waterSource = waterSource;
    }

    public void markShop() {
        _isShop = true;
    }

    public void markTavern() {
        _isTavern = true;
    }

    public boolean isShop() {
        return _isShop;
    }

    public boolean isTavern() {
        return _isTavern;
    }

    public void setDesc(String desc) {
        _desc = desc;
    }

    public String getDesc() {
        return _desc;
    }

    public void setAvailableExits(Set<Directions> availableExits) {
        _availableExits = availableExits;
    }

    public Set<Directions> getAvailableExits() {
        return _availableExits;
    }

    public void discoverMob(String mob) {
        _knownMobs.add(mob);
    }
}

File: src/main/scala/com/tort/mudai/mapper/Location.scala
package com.tort.mudai.mapper

import com.tort.mudai.RoomSnapshot
import scalaz._
import Scalaz._
import com.tort.mudai.mapper.Location.LocationId

class Location(
                val id: String @@ LocationId,
                val title: String,
                val desc: String,
                var zone: Option[Zone] = None,
                var waterSource: Option[String] = None,
                var isShop: Boolean = false,
                var isTavern: Boolean = false) {

  def markShop() {
    isShop = true
  }

  override def equals(obj: Any) = obj match {
    case location: Location => this === location
    case _ => false
  }
  override def hashCode() = id.hashCode

  def markTavern() {
    isTavern = true
  }
}

case class Exit(direction: String @@ Direction, isBorder: Boolean = false, closed: Boolean = false)

object Location {
  trait LocationId

  def locationId(id: String): String @@ LocationId = Tag(id)

  def apply(room: RoomSnapshot) = new Location(locationId("0"), room.title, room.desc)

  def apply(id: String, title: String, desc: String) = new Location(locationId(id), title, desc)

  implicit val locationsEqual: Equal[Location] = Equal.equal(_.id === _.id)
  implicit val locationIdEqual: Equal[String @@ LocationId] = Equal.equal(_ == _)
}

class Transition(
                  val id: String,
                  val from: Location,
                  val direction: String @@ Direction,
                  val to: Location,
                  val isBorder: Boolean = false,
                  val isTriggered: Boolean = false)

